README.md
# 介绍
此包是```zApibus```的Dart语言的客户端SDK。

# 使用
```dart
zApibus bus=zApibus(String appkey, String secret, String url,
      {String httpMethod = "POST",
      Map<String,dynamic> headers,
      int connectTimeout = 5000,
      int receiveTimeout = 3000});
```

## 参数说明
* appkey - 调用的授权appkey，即调用者身份
* secret - 调用的安全字符串
* url - zApibus服务器的地址
* httpMethod - 请求的方法，默认为：POST
* headers - 请求时添加的自定义http header
* connectTimeout - 连接服务器超时时间，单位是毫秒。
* receiveTimeout - 接收数据的最长时限，单位是毫秒。
* onResponse - 响应拦截函数，函数签名为：```Future Function(zApibusRequestResponse response)```
* onHeaders - 添加Header时触发，可完成动态添加Header，函数签名为：```Future<Map<String, String>> Function(Map<String, dynamic> reqData)```

> 增加```onResponse```的目的是为了拦截请求结果，判断用户是否是未登录，未登录时可统一跳转到登录页面。

### zApibusRequestResponse

在响应拦截中传入此类型数据。

* Map<String, dynamic> request - 请求的原始数据
* zApibusResponse<T> response - 解析后的响应数据
* String httpMethod - 请求的http method。目前仅支持GET/POST
* Map<String, dynamic> reqHeader - 自定义的请求头
* String url - 请求的URL
* int httpStatusCode - 响应的HTTP状态码
* String? httpResponseBody - 响应的原始数据
* Object? error - 错误信息
* StackTrace? stack - 错误的调用堆栈

## 调用方法

```dart
zApibusResponse<LoginResponse> response = await bus.execute<LoginResponse>(
      "zsea.admin.login", login,
      deserializer: Deserializer);
```
或者
```dart
  zApibusResponse<LoginResponse> response = await bus.request<LoginResponse>(
      LoginRequest("admin", "admin", "127.0.0.1"),
      deserializer: Deserializer);
```

execute方法通过api名称与```Map<String, dynamic>?```传递参数。
request方法通过继承自```zApibusRequest```类的子类传递参数。

### 其它参数说明

* authenticate - 用户认证信息。
* session - 用户的会话信息，与```authenticate```传入一个即可。
* httpMethod - 调用HTTP接口时的方法，可选值：POST/GET。
* deserializer - 调用成功时，将返回的```data```字段实例化为方法的泛型参数的回调方法。函数签名为：```T Function(dynamic d)```
* headers - 自定义的http头，与实例化时传入的进行合并。
* connectTimeout - 连接走超时时间
* receiveTimeout - 接收数据超时时间


### 返回值

两个方法均返回一个```zApibusResponse<T>```对象，其中泛型为实例化后的具体对象。

#### zApibusResponse

该对象为调用的返回结果：
* id - 当前请求的ID，一般用于日志检查
* code - 响应代码，为0时表示成功。
* message - 请求代码的描述。
* sub_code - 子错误码，一般为更详细的错误代码。
* sub_message - 子错误描述。
* data - 请求的响应结果
* origin - 请求成功的原始数据


CHANGELOG.md
## 1.3.0

* 增加```onResponse```拦截

## 1.3.1

* 为响应拦截增加更全的数据
* 修复响应数据解析无data数据时数据解析失败
* 增加```onHeaders```处理方法，可在每次请求前动态添加Header



test/zapibus_test.dart
import 'package:zapibus/src/zapibus.dart';
import 'package:crypto/crypto.dart' as crypto;
import 'dart:convert';

String _md5(String s) {
  var content = new Utf8Encoder().convert(s);
  var md5 = crypto.md5;
  var digest = md5.convert(content);
  return digest.toString();
}

class Token {
  String? token;
  String? secret;
}

class LoginResponse {
  Token? token;
  Token? refresh;
}

LoginResponse Deserializer(Map<String, dynamic> json) {
  LoginResponse response = LoginResponse();
  response.token = Token();
  response.token!.token = json["token"]["token"];
  response.token!.secret = json["token"]["secret"];
  response.refresh = Token();
  response.refresh!.token = json["refresh"]["token"];
  response.refresh!.secret = json["refresh"]["secret"];
  return response;
}

class LoginRequest extends zApibusRequest {
  String username, password, domain;
  LoginRequest(this.username, this.password, this.domain)
      : super("zsea.admin.login");
  @override
  Map<String, dynamic> toParams() {
    var timestamp = DateTime.now().microsecondsSinceEpoch ~/ 1000;
    return {
      "username": "admin",
      "timestamp": timestamp,
      "domain": "",
      "password": _md5(_md5("admin#admin") + "#" + timestamp.toString())
    };
  }
}

class MessageListResponse {
  static MessageListResponse deserializer(dynamic json) {
    MessageListResponse response = MessageListResponse();
    //response.count=json["count"];
    List<dynamic> _json = json;
    print("打印数据");
    print(_json.length);
    return response;
  }
}

void main() async {
  // test('adds one to input values', () {
  //   final calculator = Calculator();
  //   expect(calculator.addOne(2), 3);
  //   expect(calculator.addOne(-7), -6);
  //   expect(calculator.addOne(0), 1);
  // });
  var timestamp = DateTime.now().microsecondsSinceEpoch ~/ 1000;
  print("时间：$timestamp");
  Map<String, dynamic> login = {
    "username": "admin",
    "timestamp": timestamp,
    "domain": "expert.zsea.app",
    "password": _md5(_md5("admin#admin") + "#" + timestamp.toString())
  };
  zApibus bus = new zApibus("1", "hmJ1Y7tyi7cM", "http://127.0.0.1:8080/api",
      headers: {"x-token": "aaaa"});
  bus.onResponse = (response) async {
    print(response.httpResponseBody);
    //print(response.response.message);
  };
  //bus.url="";
  // zApibusResponse<LoginResponse> response = await bus.Execute<LoginResponse>(
  //     "zsea.admin.login", login,
  //     deserializer: Deserializer);
  zApibusResponse<MessageListResponse> response =
      await bus.execute<MessageListResponse>(
          "charge.user.messages.get", {"laster": 999999},
          deserializer: MessageListResponse.deserializer,
          session: "JSt7C6hmnGyM4iNArSn6JkktHBRMd96a");
  print(response);
}


lib/src/zapibus.dart
import 'dart:convert';
import 'package:crypto/crypto.dart' as crypto;
import 'package:dio/dio.dart';

String _md5(String s) {
  var content = const Utf8Encoder().convert(s);
  var md5 = crypto.md5;
  var digest = md5.convert(content);
  return digest.toString();
}

void _appendTosortByArray(
    List data, List<String> sort, String parent, String key) {
  if (data == null || data.length == 0) return;

  for (int i = 0; i < data.length; i++) {
    var value = data[i];
    if (value == null || value == "") continue;
    if (value is List) {
      _appendTosortByArray(value, sort, "$parent$key[$i]", "");
    } else if (value is Map<String, dynamic>) {
      _appendTosortByObject(value, sort, "$parent$key[$i]");
    } else {
      sort.add("$parent$key[$i]=$value");
    }
  }
}

void _appendTosortByObject(
    Map<String, dynamic> data, List<String> sort, String parent) {
  if (data == null) return;
  data.forEach((key, value) {
    if (value == null || value == "") return;
    if (value is List) {
      _appendTosortByArray(value, sort, parent, key);
    } else if (value is Map<String, dynamic>) {
      _appendTosortByObject(value, sort, parent + key + ".");
    } else {
      sort.add(parent + key + "=" + value.toString());
    }
  });
}

String _GetSignature(Map<String, dynamic> reqData, String secret) {
  List<String> s = [];
  reqData.forEach((key, value) {
    if (key == "signature" || value == null || value == "") return;
    if (key == "request") {
      if (value is Map<String, dynamic>) {
        _appendTosortByObject(value, s, "request.");
      }
      return;
    }
    s.add("$key=$value");
  });
  s.add("secret=" + secret);
  s.sort();
  var plain = s.join("&");
  var signature = _md5(plain);
  return signature;
}

abstract class zApibusRequest {
  String method = "";
  zApibusRequest(this.method);
  Map<String, dynamic> toParams();
}

class zApibusAuthenticate {
  String token;
  String secret;
  zApibusAuthenticate(this.token, this.secret);
}

class zApibusResponse<T> {
  String id = "";
  int code = 0;
  String? sub_code = "";
  String message = "";
  String? sub_message = "";
  T? data;
  dynamic origin;
  zApibusResponse(
      this.id, this.code, this.message, this.sub_code, this.sub_message);
  factory zApibusResponse.fromJson(
      Map<String, dynamic> json, T Function(dynamic d)? deserializer) {
    zApibusResponse<T> response = zApibusResponse(json['id'], json['code'],
        json['message'], json['sub_code'], json['sub_message']);
    if (deserializer != null && json["data"] != null) {
      response.data = deserializer(json["data"]);
    }
    response.origin = json["data"];
    return response;
  }
}

class zApibusRequestResponse<T> {
  Map<String, dynamic> request;
  zApibusResponse<T> response;
  String httpMethod;
  Map<String, dynamic> reqHeader;
  String url;
  int httpStatusCode;
  String? httpResponseBody;
  Object? error;
  StackTrace? stack;
  zApibusRequestResponse(
      {required this.request,
      required this.response,
      required this.httpMethod,
      required this.reqHeader,
      required this.url,
      required this.httpStatusCode,
      this.httpResponseBody,
      this.error,
      this.stack});
}

class zApibus {
  String appkey, secret, url, httpMethod = "POST";
  Map<String, String>? headers;
  Future Function(zApibusRequestResponse response)? onResponse;
  Future<Map<String, String>> Function(Map<String, dynamic> reqData)? onHeaders;
  final Dio _dio = Dio();
  zApibus(this.appkey, this.secret, this.url,
      {this.httpMethod = "POST",
      this.headers,
      int connectTimeout = 5000,
      int receiveTimeout = 3000,
      this.onResponse,
      this.onHeaders}) {
    _dio.options.connectTimeout = connectTimeout;
    _dio.options.receiveTimeout = receiveTimeout;
  }
  Future<zApibusResponse<T>> execute<T>(
      String method, Map<String, dynamic>? params,
      {zApibusAuthenticate? authenticate,
      String? session,
      String? httpMethod,
      T Function(dynamic d)? deserializer,
      Map<String, dynamic>? headers,
      int? connectTimeout,
      int? receiveTimeout}) async {
    zApibusRequestResponse<T> response = await _execute<T>(method, params,
        authenticate: authenticate,
        session: session,
        httpMethod: httpMethod,
        deserializer: deserializer,
        headers: headers,
        connectTimeout: connectTimeout,
        receiveTimeout: receiveTimeout);
    if (onResponse != null) {
      await onResponse!(response);
    }
    return response.response;
  }

  Future<zApibusRequestResponse<T>> _execute<T>(
      String method, Map<String, dynamic>? params,
      {zApibusAuthenticate? authenticate,
      String? session,
      String? httpMethod,
      T Function(dynamic d)? deserializer,
      Map<String, dynamic>? headers,
      int? connectTimeout,
      int? receiveTimeout}) async {
    //options.noSuchMethod(invocation)
    Map<String, dynamic> reqData = {
      'appkey': appkey,
      'time': DateTime.now().microsecondsSinceEpoch ~/ (1000 * 1000),
      'method': method,
      'request': params,
      'session': session,
      'token': authenticate?.token
    };
    String signature = _GetSignature(
        reqData, authenticate == null ? secret : authenticate.secret);
    reqData.addAll({"signature": signature});
    Response<dynamic> response;
    String http_method = httpMethod ?? this.httpMethod;
    Map<String, dynamic> __headers = {};
    if (this.headers != null) {
      __headers.addAll(this.headers!);
    }
    if (headers != null) {
      __headers.addAll(headers);
    }
    if (onHeaders != null) {
      var _xHeaders = await onHeaders!(reqData);
      __headers.addAll(_xHeaders);
    }
    try {
      if (http_method == "GET") {
        response = await _dio.get(
          url,
          queryParameters: {
            "body": reqData,
            "_": DateTime.now().microsecondsSinceEpoch
          },
          options: Options(
            headers: __headers,
          ),
        );
      } else {
        response = await _dio.post(
          url,
          data: reqData,
          options: Options(
            headers: __headers,
          ),
        );
      }
    } catch (e, stack) {
      return zApibusRequestResponse(
          request: reqData,
          response: zApibusResponse("", 10, "Service Currently Unavailable",
              "isp.apibus-unknown-error", "请求zApibus时发生未知错误"),
          httpMethod: http_method,
          reqHeader: __headers,
          url: url,
          httpStatusCode: 0,
          error: e,
          stack: stack);
    }
    if (response.statusCode != 200) {
      return zApibusRequestResponse(
          request: reqData,
          response: zApibusResponse(
              "",
              10,
              "Service Currently Unavailable",
              "isp.apibus-net-error:httpcode:" + response.statusCode.toString(),
              "zApibus服务器错误"),
          httpMethod: http_method,
          reqHeader: __headers,
          url: url,
          httpStatusCode: response.statusCode ?? 0,
          httpResponseBody: response.data);
    }
    //print(response.data.runtimeType);
    JsonCodec JSON = const JsonCodec();
    try {
      zApibusResponse<T> zResponse =
          zApibusResponse.fromJson(JSON.decode(response.data), deserializer);

      return zApibusRequestResponse(
        request: reqData,
        response: zResponse,
        httpMethod: http_method,
        reqHeader: __headers,
        url: url,
        httpStatusCode: response.statusCode ?? 0,
        httpResponseBody: response.data,
      );
    } catch (e, stack) {
      return zApibusRequestResponse(
          request: reqData,
          response: zApibusResponse("", 10, "Service Currently Unavailable",
              "isp.apibus-response-format-error", "zApibus服务器响应数据格式错误"),
          httpMethod: http_method,
          reqHeader: __headers,
          url: url,
          httpStatusCode: 0,
          httpResponseBody: response.data,
          error: e,
          stack: stack);
    }
  }

  Future<zApibusResponse<T>> request<T>(zApibusRequest request,
      {zApibusAuthenticate? authenticate,
      String? session,
      String? httpMethod,
      T Function(dynamic d)? deserializer,
      Map<String, dynamic>? headers,
      int? connectTimeout,
      int? receiveTimeout}) async {
    return execute<T>(request.method, request.toParams(),
        authenticate: authenticate,
        session: session,
        httpMethod: httpMethod,
        deserializer: deserializer,
        headers: headers,
        connectTimeout: connectTimeout,
        receiveTimeout: receiveTimeout);
  }
}
